File: project/.sdk/tm/php/utility/MakePoint.php
<?php
declare(strict_types=1);

// ProjectName SDK utility: make_point

require_once __DIR__ . '/../core/Helpers.php';

class ProjectNameMakePoint
{
    public static function call(ProjectNameContext $ctx): array
    {
        if (isset($ctx->out['point'])) {
            $ctx->point = $ctx->out['point'];
            return [$ctx->point, null];
        }

        $op = $ctx->op;
        $options = $ctx->options;

        $allow_op = \Voxgig\Struct\Struct::getpath($options, 'allow.op') ?? '';
        if (strpos($allow_op, $op->name) === false) {
            return [null, $ctx->make_error('point_op_allow',
                "Operation \"{$op->name}\" not allowed by SDK option allow.op value: \"{$allow_op}\"")];
        }

        if (empty($op->points)) {
            return [null, $ctx->make_error('point_no_points',
                "Operation \"{$op->name}\" has no endpoint definitions.")];
        }

        if (count($op->points) === 1) {
            $ctx->point = $op->points[0];
        } else {
            $reqselector = $op->input === 'data' ? $ctx->reqdata : $ctx->reqmatch;
            $selector = $op->input === 'data' ? $ctx->data : $ctx->match;

            $point = null;
            foreach ($op->points as $p) {
                $point = $p;
                $select_def = ProjectNameHelpers::to_map(\Voxgig\Struct\Struct::getprop($p, 'select'));
                $found = true;

                if ($selector && $select_def) {
                    $exist = \Voxgig\Struct\Struct::getprop($select_def, 'exist');
                    if (is_array($exist)) {
                        foreach ($exist as $ek) {
                            $rv = \Voxgig\Struct\Struct::getprop($reqselector, (string)$ek);
                            $sv = \Voxgig\Struct\Struct::getprop($selector, (string)$ek);
                            if ($rv === null && $sv === null) {
                                $found = false;
                                break;
                            }
                        }
                    }
                }

                if ($found) {
                    $req_action = \Voxgig\Struct\Struct::getprop($reqselector, '$action');
                    $select_action = \Voxgig\Struct\Struct::getprop($select_def, '$action');
                    if ($req_action !== $select_action) {
                        $found = false;
                    }
                }

                if ($found) {
                    break;
                }
            }

            if ($reqselector) {
                $req_action = \Voxgig\Struct\Struct::getprop($reqselector, '$action');
                if ($req_action && $point) {
                    $point_select = ProjectNameHelpers::to_map(\Voxgig\Struct\Struct::getprop($point, 'select'));
                    $point_action = \Voxgig\Struct\Struct::getprop($point_select, '$action');
                    if ($req_action !== $point_action) {
                        return [null, $ctx->make_error('point_action_invalid',
                            "Operation \"{$op->name}\" action \"" . \Voxgig\Struct\Struct::stringify($req_action) . "\" is not valid.")];
                    }
                }
            }

            $ctx->point = $point;
        }

        return [$ctx->point, null];
    }
}

File: project/.sdk/tm/go/utility/make_point.go
package utility

import (
	"strings"

	vs "github.com/voxgig/struct"

	"GOMODULE/core"
)

func makePointUtil(ctx *core.Context) (map[string]any, error) {
	if ctx.Out["point"] != nil {
		if tm, ok := ctx.Out["point"].(map[string]any); ok {
			ctx.Point = tm
			return tm, nil
		}
	}

	op := ctx.Op
	options := ctx.Options

	allowOp, _ := vs.GetPath([]any{"allow", "op"}, options).(string)
	if !strings.Contains(allowOp, op.Name) {
		return nil, ctx.MakeError("point_op_allow",
			"Operation \""+op.Name+
				"\" not allowed by SDK option allow.op value: \""+allowOp+"\"")
	}

	if len(op.Points) == 0 {
		return nil, ctx.MakeError("point_no_points",
			"Operation \""+op.Name+"\" has no endpoint definitions.")
	}

	if len(op.Points) == 1 {
		ctx.Point = op.Points[0]
	} else {
		var reqselector map[string]any
		var selector map[string]any

		if op.Input == "data" {
			reqselector = ctx.Reqdata
			selector = ctx.Data
		} else {
			reqselector = ctx.Reqmatch
			selector = ctx.Match
		}

		var point map[string]any
		for i := 0; i < len(op.Points); i++ {
			point = op.Points[i]
			selectDef := core.ToMapAny(vs.GetProp(point, "select"))
			found := true

			if selector != nil && selectDef != nil {
				if exist := vs.GetProp(selectDef, "exist"); exist != nil {
					if existList, ok := exist.([]any); ok {
						for _, ek := range existList {
							existkey, _ := ek.(string)
							rv := vs.GetProp(reqselector, existkey)
							sv := vs.GetProp(selector, existkey)
							if rv == nil && sv == nil {
								found = false
								break
							}
						}
					}
				}
			}

			if found {
				reqAction := vs.GetProp(reqselector, "$action")
				selectAction := vs.GetProp(selectDef, "$action")
				if reqAction != selectAction {
					found = false
				}
			}

			if found {
				break
			}
		}

		if reqselector != nil {
			reqAction := vs.GetProp(reqselector, "$action")
			if reqAction != nil && point != nil {
				pointSelect := core.ToMapAny(vs.GetProp(point, "select"))
				pointAction := vs.GetProp(pointSelect, "$action")
				if reqAction != pointAction {
					return nil, ctx.MakeError("point_action_invalid",
						"Operation \""+op.Name+
							"\" action \""+vs.Stringify(reqAction)+"\" is not valid.")
				}
			}
		}

		ctx.Point = point
	}

	return ctx.Point, nil
}
